### PR TITLE
more efficient sample op

### PR DESCRIPTION
**Pull request by rawls238**
@eytan

Another thing that I found was rather inefficient was the sample op since if we're only sampling 5 elements from a 1000 element population it doesn't make sense from an efficiency standpoint to have to iterate through all 1000 of them.

Due to the way that the Fisher-Yates algorithm works, we know that for if we are on iteration i, then the elements for n - i + 1 are fixed.  We can still have the same desirable properties from a statistical perspective (i.e. every element has an equal probability of being in the sample + in any arbitrary position in the sample - I did some quick calculations to verify this + the sample op tests still pass but I would appreciate some double checking of this) with our sample op if we early return once we have k elements (where k is the number of draws) that are fixed. As the segment size increases this can make a real impact performance-wise.

Running the namespace initialization benchmarks gets the following results now (which are a slightly skewed test for this since they only allocate 5 segments for each added experiment, but the differences are still meaningful):

Result w/ compatibility turned on: 146.8896060003899 ms over 500 runs
Result w/ compatibility turned off (without this sampling change): 19.26401599845849 ms over 500 runs
Result w/ compatibility turned off + this change: 0.4321100015658885 ms over 500 runs

In the worst case this is just as bad as without the changes, but for the average case it's much better.

This of course breaks compatibility with other implementations + previous versions of the JS implementation, but we should consider changing this in the ref implementation.

### CODE REVIEW

**Comment by eytan** [comment]
This sounds awesome!  I'll try to take a closer look soon, and also verify that the unit test is doing the right thing... In the meantime, do you have a proof that this works correctly?

**Comment by rawls238** [comment]
@eytan yea I'll try to write it out tomorrow to verify (or disprove :)) its correctness tmw- I'm currently on a train with 5% battery left on my laptop.

File: es6/ops/random.js [hunk 1]
```diff
@@ -160,8 +160,12 @@ class WeightedChoice extends PlanOutOpRandom {
...
-  shuffle(array) {
-    for (var i = array.length - 1; i > 0; i--) {
+  sample(array, numDraws) {
+    var len = array.length;
+    var stoppingPoint = len - numDraws;
+    var c = usingCompatibleHash();
+
+    for (var i = len - 1; i > 0; i--) {
       var j;
       if (usingCompatibleHash()) {
```
Comment (TimBeyer) [inline es6/ops/random.js]: Could make this also use `c`

**Comment by TimBeyer** [comment]
Really nice work. Since Fisher-Yates only ever swaps elements at a smaller or equal index, as you said, there should be no difference in the distribution of samples. This is a great optimization! :)

I'm starting to wonder though if it might be worth thinking about a better abstraction for the compatibility vs non-compatibility mode. Right now it's still quite readable, and the hashing implementation doesn't really leak outside of the random ops module, but maybe it will make the verification of the statistics involved, and producing optimized bundles simpler. I got some free time on my hands, so maybe I can come up with a proposal.

File: es6/ops/random.js (hunk 1 above)
Comment (rawls238) [inline es6/ops/random.js]: :+1: good catch

**Comment by rawls238** [comment]
Quick proof of correctness for this:

Effectively a sample is a random permutation of size k which means that there are ```n! / (n-k)!``` samples for a sample size of k. All we have to show is that with early return from the fisher-yates shuffle each permutation has probability ```(n-k)! / n!``` to occur.

We'll pick an arbitrary permutation of k elements and show that it has ```(n-k)! / n!``` probability of being the selected permutation by the algorithm. The first position in the permutation has probability ```1 / n``` of being selected. After this step, the element is fixed in that position and we select from the remaining ```n - 1``` options. The second position in the permutation has probability ```1 / n -1``` of being selected, etc. up until the kth element in the permutation which has probability ```1 / n - k + 1``` probability of being selected. This means that the probability of this arbitrary permutation being selected after k iterations of the algorithm is ``` (1 / n) * (1 / n - 1) * (1 / n - 2) * ... * (1 / (n - k + 1)) = (n-k)! / n!```. Since the selection of the original permutation was arbitrary this allows to all k permutations which is precisely what we want.

**Comment by rawls238** [comment]
@TimBeyer would be happy if you came up with a proposal. One thing that's important (at least from my perspective) is having a way to slowly transition existing namespaces / experiments onto using the non-compatible version as it would become possible to transition them (i.e. their segment size would hit 0). I'm not sure entirely certain that the current compatibility toggle supports that well, but I think that including multiple bundles may make such efforts unnecessarily difficult and costly. Let me know if you can think of a good solution to the issue.

**Comment by rawls238** [comment]
I'm leaning towards just turning compatibility off by default and having it be opt-in. If other people complain that they need a more clear upgrade path for namespace by namespace then I think it's worth investing time in, but I think it's unneeded unless explicitly requested. @TimBeyer would you be fine with that?

**Comment by TimBeyer** [comment]
I believe this would be fine for us since assignments are persisted in cookies and reassignments happen automatically once experiments are added or concluded. So current membership would stay the same, new membership would be assigned according to the new hashing, following the same distribution, and once experiments change everyone would be assigned using the new hashing.

**Comment by eytan** [comment]
Hi @rawls238 -- this all looks good: the modification seems straightforward and solid, and this is such a huge performance improvement that compatibility mode should definitely be off by default.  One thing to note though is that the original sample() unit test is not particularly rigorous: it only tests to make sure that, for example, given 1, 2, 3 draws from a list of elements yields their expected probabilities; it does not actually check whether the expected subsequences occur with equal probability.

**Comment by rawls238** [comment]
:+1: going to merge this and then add some docs and release a new version.

**Pull request merged**